pg_wait_sampling.c

```diff
@@ -555,129 +485,45 @@ pg_wait_sampling_get_current(PG_FUNCTION_ARGS)
-	recv_mq = shm_mq_create(pgws_collector_mq, COLLECTOR_QUEUE_SIZE);
-	pgws_collector_hdr->request = request;
-	LockRelease(&collectorTag, ExclusiveLock, false);
-
-	if (!pgws_collector_hdr->latch)
-		ereport(ERROR, (errcode(ERRCODE_INTERNAL_ERROR),
-						errmsg("pg_wait_sampling collector wasn't started")));
-
-	SetLatch(pgws_collector_hdr->latch);
-
-	shm_mq_set_receiver(recv_mq, MyProc);
-
-	/*
-	 * We switch to TopMemoryContext, so that recv_mqh is allocated there
-	 * and is guaranteed to survive until before_shmem_exit callbacks are
-	 * fired.  Anyway, shm_mq_detach() will free handler on its own.
-	 *
-	 * NB: we do not pass `seg` to shm_mq_attach(), so it won't set its own
-	 * callback, i.e. we do not interfere here with shm_mq_detach_callback().
-	 */
-	oldctx = MemoryContextSwitchTo(TopMemoryContext);
-	recv_mqh = shm_mq_attach(recv_mq, NULL, NULL);
-	MemoryContextSwitchTo(oldctx);
-
-	/*
-	 * Now we surely attached to the shm_mq and got collector's attention.
-	 * If anything went wrong (e.g. Ctrl+C received from the client) we have
-	 * to cleanup some things, i.e. detach from the shm_mq, so collector was
-	 * able to continue responding to other requests.
-	 *
-	 * PG_ENSURE_ERROR_CLEANUP() guaranties that cleanup callback will be
-	 * fired for both ERROR and FATAL.
-	 */
-	PG_ENSURE_ERROR_CLEANUP(pgws_cleanup_callback, 0);
-	{
-		res = shm_mq_receive(recv_mqh, &len, &data, false);
-		if (res != SHM_MQ_SUCCESS || len != sizeof(*count))
-			elog(ERROR, "error reading mq");
-
-		memcpy(count, data, sizeof(*count));
-
-		result = palloc(item_size * (*count));
-		ptr = result;
-
-		for (i = 0; i < *count; i++)
-		{
-			res = shm_mq_receive(recv_mqh, &len, &data, false);
-			if (res != SHM_MQ_SUCCESS || len != item_size)
-				elog(ERROR, "error reading mq");
-
-			memcpy(ptr, data, item_size);
-			ptr += item_size;
-		}
-	}
-	PG_END_ENSURE_ERROR_CLEANUP(pgws_cleanup_callback, 0);
-
-	/* We still have to detach and release lock during normal operation. */
-	shm_mq_detach_compat(recv_mqh, recv_mq);
-	LockRelease(&queueTag, ExclusiveLock, false);
-
-	return result;
-}
-
 
 PG_FUNCTION_INFO_V1(pg_wait_sampling_get_profile);
 Datum
 pg_wait_sampling_get_profile(PG_FUNCTION_ARGS)
 {
-	Profile			   *profile;
-	FuncCallContext	   *funcctx;
+	ProfileHashEntry	*profile;
+	FuncCallContext		*funcctx;
 
 	check_shmem();
 
 	if (SRF_IS_FIRSTCALL())
 	{
 		MemoryContext		oldcontext;
 		TupleDesc			tupdesc;
+		HASH_SEQ_STATUS		hash_seq;
+		ProfileHashEntry   *entry;
+		int					profile_count,
+							entry_index;
 
 		funcctx = SRF_FIRSTCALL_INIT();
 		oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);
 
-		/* Receive profile from shmq */
-		profile = (Profile *) palloc0(sizeof(Profile));
-		profile->items = (ProfileItem *) receive_array(PROFILE_REQUEST,
-										sizeof(ProfileItem), &profile->count);
+		/* Extract profile from shared memory */
+		profile_count = hash_get_num_entries(pgws_profile_hash);
```

Comment:
We should do this after locking the hash. Otherwise, we would allocate not enough memory if the hash grows before we lock it.

Reply:
Yes, you are right. Thanks